private/model/api/eventstream.go

```diff
@@ -516,22 +622,443 @@ func (s *{{ $.ShapeName }}) UnmarshalEvent(
+		{{ end }}
+	{{ end }}
+{{ end }}
+
+type loopReader struct {
+	source *bytes.Reader
+}
+
+func (c *loopReader) Read(p []byte) (int, error) {
+	if c.source.Len() == 0 {
+		c.source.Seek(0, 0)
+	}
+
+	return c.source.Read(p)
+}
+
+{{ define "event stream inbound tests" }}
+	func Test{{ $.Operation.ExportedName }}_Read(t *testing.T) {
+		expectEvents, eventMsgs := mock{{ $.Operation.ExportedName }}ReadEvents()
+		sess, cleanupFn, err := eventstreamtest.SetupEventStreamSession(t,
+			eventstreamtest.ServeEventStream{
+				T:      t,
+				Events: eventMsgs,
+			},
+			true,
+		)
+		if err != nil {
+			t.Fatalf("expect no error, %v", err)
+		}
+		defer cleanupFn()
+
+		svc := New(sess)
+		resp, err := svc.{{ $.Operation.ExportedName }}(nil)
+		if err != nil {
+			t.Fatalf("expect no error got, %v", err)
+		}
+		defer resp.EventStream.Close()
+
+		{{- if eq $.Operation.API.Metadata.Protocol "json" }}
+			expectResp := expectEvents[0].(*{{ $.Operation.OutputRef.Shape.ShapeName }})
+			{{- range $name, $ref := $.Operation.OutputRef.Shape.MemberRefs }}
+				{{- if not $ref.Shape.IsEventStream }}
+					if e, a := expectResp.{{ $name }}, resp.{{ $name }}; !reflect.DeepEqual(e,a) {
+						t.Errorf("expect %v, got %v", e, a)
+					}
+				{{- end }}
+			{{- end }}
+			expectEvents = expectEvents[1:]
+		{{ end }}
+
+		var i int
+		for event := range resp.EventStream.Events() {
+			if event == nil {
+				t.Errorf("%d, expect event, got nil", i)
+			}
+			if e, a := expectEvents[i], event; !reflect.DeepEqual(e, a) {
+				t.Errorf("%d, expect %T %v, got %T %v", i, e, e, a, a)
+			}
+			i++
+		}
+
+		if err := resp.EventStream.Err(); err != nil {
+			t.Errorf("expect no error, %v", err)
+		}
+	}
+
+	func Test{{ $.Operation.ExportedName }}_ReadClose(t *testing.T) {
+		_, eventMsgs := mock{{ $.Operation.ExportedName }}ReadEvents()
+		sess, cleanupFn, err := eventstreamtest.SetupEventStreamSession(t,
+			eventstreamtest.ServeEventStream{
+				T:      t,
+				Events: eventMsgs,
+			},
+			true,
+		)
+		if err != nil {
+			t.Fatalf("expect no error, %v", err)
+		}
+		defer cleanupFn()
+
+		svc := New(sess)
+		resp, err := svc.{{ $.Operation.ExportedName }}(nil)
+		if err != nil {
+			t.Fatalf("expect no error got, %v", err)
+		}
+
+		resp.EventStream.Close()
+
+		if err := resp.EventStream.Err(); err != nil {
+			t.Errorf("expect no error, %v", err)
+		}
+	}
+
+	func Benchmark{{ $.Operation.ExportedName }}_Read(b *testing.B) {
+		_, eventMsgs := mock{{ $.Operation.ExportedName }}ReadEvents()
+		var buf bytes.Buffer
+		encoder := eventstream.NewEncoder(&buf)
+		for _, msg := range eventMsgs {
+			if err := encoder.Encode(msg); err != nil {
+				b.Fatalf("failed to encode message, %v", err)
+			}
+		}
+		stream := &loopReader{source: bytes.NewReader(buf.Bytes())}
+
+		sess := unit.Session
+		svc := New(sess, &aws.Config{
+			Endpoint:               aws.String("https://example.com"),
+			DisableParamValidation: aws.Bool(true),
```

Comment:
Why is this disabled for benchmarks but not unit tests?

Reply:
Forgot tests do actually disable the parameter validation in the `eventstreamtest.SetupEventStreamSession` helper.